### PR TITLE
gcc-10 (parpack) and mpich

### LINKED ISSUE #241: Some tests fail with mpich
## Expected behavior

All tests pass

## Actual behavior

90/90 Test #90: icb_parpack_cpp_tst ..............   Passed    0.16 sec

91% tests passed, 8 tests failed out of 90

Total Test time (real) =   1.54 sec

The following tests FAILED:
         72 - pcndrv1_ex (Failed)
         73 - pdndrv1_ex (Failed)
         74 - pdndrv3_ex (Failed)
         75 - pdsdrv1_ex (Failed)
         76 - psndrv1_ex (Failed)
         77 - psndrv3_ex (Failed)
         78 - pssdrv1_ex (Failed)
         79 - pzndrv1_ex (Failed)
Errors while running CTest

## Where/how to reproduce the problem

- arpack-ng: master (6b04aa4bcce47cbf51579314cb099062cc716c86)
- OS: fedora 30, fedora 33
- compiler: gcc 9 and gcc 10 (with fixes for gcc 10)
- environment: module load mpi/mpich-x86_64
- configure: cmake -DEXAMPLES=ON -DMPI=ON -DICB=ON .. && make -j 4 && make test

## Steps to reproduce the problem
```
docker run fedora
dnf -y install emacs git autoconf libtool make gcc-c++ gfortran openblas-devel environment-modules mpich-devel
. /etc/profile.d/modules.sh
module load mpi/mpich-x86_64
cmake -DEXAMPLES=ON -DMPI=ON -DICB=ON .. && make -j 4 && make test
```

## Error message

see above

## Traces

```
 --------------------------------------------------------
    1 -    2: (-5.04608E+01,-3.56467E+02)  ( 9.82758E+01, 3.43965E+02)
    3 -    3: ( 7.62495E-04, 0.00000E+00)


 _naup2: no. of "converged" Ritz values at this iter.
 ----------------------------------------------------
    1 -    1:       0


 _napps: matrix splitting at row/column no.
 ------------------------------------------
    1 -    1:       2


 _napps: matrix splitting with shift number.
 -------------------------------------------
    1 -    1:       2


 _napps: off diagonal element.
 -----------------------------
    1 -    1: (-6.10352E-05, 3.05176E-05)


 _naup2: **** Start of major iteration number ****
 -------------------------------------------------
    1 -    1:    1015


 _ngets: KEV is
 --------------
    1 -    1:       1


 _ngets: NP is
 -------------
    1 -    1:       2


 _ngets: Eigenvalues of current H matrix
 ----------------------------------------
    1 -    2: ( 1.43426E+02, 1.43426E+02)  ( 8.61754E+02, 8.61754E+02)
    3 -    3: ( 1.00004E+03, 1.00004E+03)


 _ngets: Ritz estimates of the current KEV+NP Ritz values
 --------------------------------------------------------
    1 -    2: (-1.00379E+02,-3.32978E+02)  (-3.44731E+02, 5.06945E-10)
    3 -    3: ( 8.00448E-04, 0.00000E+00)


 _naup2: no. of "converged" Ritz values at this iter.
 ----------------------------------------------------
    1 -    1:       0


 _napps: matrix splitting at row/column no.
 ------------------------------------------
    1 -    1:       2


 _napps: matrix splitting with shift number.
 -------------------------------------------
    1 -    1:       2


 _napps: off diagonal element.
 -----------------------------
    1 -    1: ( 1.06812E-04, 2.48815E-05)


 _naup2: **** Start of major iteration number ****
 -------------------------------------------------
    1 -    1:    1016


 _ngets: KEV is
 --------------
    1 -    1:       1


 _ngets: NP is
 -------------
    1 -    1:       2


 _ngets: Eigenvalues of current H matrix
 ----------------------------------------
    1 -    2: ( 8.43747E+02, 8.43747E+02)  ( 1.50688E+02, 1.50688E+02)
    3 -    3: ( 1.00004E+03, 1.00004E+03)


 _ngets: Ritz estimates of the current KEV+NP Ritz values
 --------------------------------------------------------
    1 -    2: (-5.04610E+01,-3.56466E+02)  (-1.98433E+02,-2.97650E+02)
    3 -    3: ( 7.49836E-04, 0.00000E+00)


 _naup2: no. of "converged" Ritz values at this iter.
 ----------------------------------------------------
    1 -    1:       0


 _napps: matrix splitting at row/column no.
 ------------------------------------------
    1 -    1:       2


 _napps: matrix splitting with shift number.
 -------------------------------------------
    1 -    1:       2


 _napps: off diagonal element.
 -----------------------------
    1 -    1: (-6.10352E-05, 3.05176E-05)


 _naup2: **** Start of major iteration number ****
 -------------------------------------------------
    1 -    1:    1017


 _ngets: KEV is
 --------------
    1 -    1:       1


 _ngets: NP is
 -------------
    1 -    1:       2


 _ngets: Eigenvalues of current H matrix
 ----------------------------------------
    1 -    2: ( 1.43426E+02, 1.43426E+02)  ( 8.61754E+02, 8.61754E+02)
    3 -    3: ( 1.00004E+03, 1.00004E+03)


 _ngets: Ritz estimates of the current KEV+NP Ritz values
 --------------------------------------------------------
    1 -    2: (-2.77857E+02,-2.09155E+02)  (-4.92123E-10,-3.44731E+02)
    3 -    3: ( 7.87159E-04, 0.00000E+00)


 _naup2: no. of "converged" Ritz values at this iter.
 ----------------------------------------------------
    1 -    1:       0


 _napps: matrix splitting at row/column no.
 ------------------------------------------
    1 -    1:       2


 _napps: matrix splitting with shift number.
 -------------------------------------------
    1 -    1:       2


 _napps: off diagonal element.
 -----------------------------
    1 -    1: ( 0.00000E+00, 4.01403E-05)


 _naup2: **** Start of major iteration number ****
 -------------------------------------------------
    1 -    1:    1018


 _ngets: KEV is
 --------------
    1 -    1:       1


 _ngets: NP is
 -------------
    1 -    1:       2


 _ngets: Eigenvalues of current H matrix
 ----------------------------------------
    1 -    2: ( 8.43746E+02, 8.43746E+02)  ( 1.50689E+02, 1.50689E+02)
    3 -    3: ( 1.00004E+03, 1.00004E+03)


 _ngets: Ritz estimates of the current KEV+NP Ritz values
 --------------------------------------------------------
    1 -    2: ( 2.25260E+02,-2.80842E+02)  ( 2.97651E+02,-1.98434E+02)
    3 -    3: ( 7.37387E-04, 1.87069E-10)


 _naup2: no. of "converged" Ritz values at this iter.
 ----------------------------------------------------
    1 -    1:       0


 _napps: matrix splitting at row/column no.
 ------------------------------------------
    1 -    1:       2


 _napps: matrix splitting with shift number.
 -------------------------------------------
    1 -    1:       2


 _napps: off diagonal element.
 -----------------------------
    1 -    1: ( 0.00000E+00, 6.91544E-06)


 _naup2: **** Start of major iteration number ****
 -------------------------------------------------
    1 -    1:    1019


 _ngets: KEV is
 --------------
    1 -    1:       1


 _ngets: NP is
 -------------
    1 -    1:       2


 _ngets: Eigenvalues of current H matrix
 ----------------------------------------
    1 -    2: ( 1.43426E+02, 1.43426E+02)  ( 8.61754E+02, 8.61754E+02)
    3 -    3: ( 1.00004E+03, 1.00004E+03)


 _ngets: Ritz estimates of the current KEV+NP Ritz values
 --------------------------------------------------------
    1 -    2: (-9.85221E+01,-3.33534E+02)  (-3.44731E+02, 0.00000E+00)
    3 -    3: ( 0.00000E+00, 0.00000E+00)


 _naupd: Number of update iterations taken
 -----------------------------------------
    1 -    1:    1019


 _naupd: Number of wanted "converged" Ritz values
 ------------------------------------------------
    1 -    1:       1


 _naupd: The final Ritz values
 -----------------------------
    1 -    1: ( 1.00004E+03, 1.00004E+03)


 _naupd: Associated Ritz estimates
 ---------------------------------
    1 -    1: ( 0.00000E+00, 0.00000E+00)



     =============================================
     = Complex implicit Arnoldi update code      =
     = Version Number:  2.1                      =
     = Version Date:    3/19/97                 =
     =============================================
     = Summary of timing statistics              =
     =============================================


     Total number update iterations             =  1019
     Total number of OP*x operations            =  2039
     Total number of B*x operations             =     0
     Total number of reorthogonalization steps  =  2039
     Total number of iterative refinement steps =     0
     Total number of restart steps              =     0
     Total time in user OP*x operation          =     0.000000
     Total time in user B*x operation           =     0.000000
     Total time in Arnoldi update routine       =     0.000000
     Total time in p_naup2 routine              =     0.000000
     Total time in basic Arnoldi iteration loop =     0.000000
     Total time in reorthogonalization phase    =     0.000000
     Total time in (re)start vector generation  =     0.000000
     Total time in Hessenberg eig. subproblem   =     0.000000
     Total time in getting the shifts           =     0.000000
     Total time in applying the shifts          =     0.000000
     Total time in convergence testing          =     0.000000
     Total time in computing final Ritz vectors =     0.000000

rank 0rank 1 -  - 1000.04 1000.041000.04 1000.04

<end of output>
Test time =   0.16 sec
----------------------------------------------------------
Test Passed.
"icb_parpack_cpp_tst" end time: Feb 20 00:01 GMT
"icb_parpack_cpp_tst" time elapsed: 00:00:00
----------------------------------------------------------

End testing: Feb 20 00:01 GMT
```

## Callstack

n.a.

## Notes, remarks

switching to openmpi resolves the issue

### LINKED ISSUE #242: Porting to gcc 10
Follow up of #239 

As mentioned, the parallel version isn't ported yet.
I am working on it, this is just to have this recorded somewhere that this is a known issue.

### PR DESCRIPTION
**Pull request by dschwoerer**
## Pull request purpose

Fixes #241 
Fixes #242 

## Detailed changes proposed in this pull request

- Make sure we pass rank-1 vectors to MPI routines
- Make sure we pass rank-1 vectors to vector-printing routines
- Make sure we use a valid address for local binaries
- Add test with gcc-10 and mpich

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -301,6 +301,33 @@ jobs:
       &&                                                                                    \
       sudo docker start -a mobydick
 
+  # fedora (with gcc 10 and mpich)
+  - stage: fedora
+    dist: xenial
+    script: |
+      sudo docker pull registry.fedoraproject.org/fedora:rawhide                         || \
```
Comment (sylvestre) [inline .travis.yml]: This is too much duplication compared to the previous task.
Could you please move some of the work in a bash script and call with it with a bash -v

Comment (dschwoerer) [reply .travis.yml]: Should I put it in a script like `.travis_fedora.sh`? Or how can I do this directly in `.travis.yml`?

Comment (sylvestre) [reply .travis.yml]: yeah, a new file in the repo
probably more scripts/travis_fedora.sh to make it more discoverable
thanks

File: .travis.yml [hunk 2]
```diff
@@ -265,68 +265,15 @@ jobs:
...
-       cmake -DEXAMPLES=ON -DMPI=ON -DICB=ON ..                                          && \
-       export VERBOSE=1                                                                  && \
-       make all                                                                          && \
-       make test                                                                         && \
-       tail -n 300 ./Testing/Temporary/LastTest.log"                                        \
-      &&                                                                                    \
-      sudo docker cp -a ${TRAVIS_BUILD_DIR} mobydick:/tmp                                   \
-      &&                                                                                    \
-      sudo docker start -a mobydick
+    script: scripts/travis_fedora.sh setup mpich rawhide
```
Comment (fghoussen) [inline .travis.yml]: @sylvestre: endless-pain is boring-time-wasting ?! docker-run jobs are _all_ different in many slight details, no way you have one script for all !? you need that here https://github.com/opencollab/arpack-ng/blob/3dd2d5fe97ed40759b3b175d4a1c27ebcd9b0921/.travis.yml#L156 but not there https://github.com/opencollab/arpack-ng/blob/3dd2d5fe97ed40759b3b175d4a1c27ebcd9b0921/.travis.yml#L185, you need that here https://github.com/opencollab/arpack-ng/blob/3dd2d5fe97ed40759b3b175d4a1c27ebcd9b0921/.travis.yml#L159 but not there https://github.com/opencollab/arpack-ng/blob/3dd2d5fe97ed40759b3b175d4a1c27ebcd9b0921/.travis.yml#L187, you need that here https://github.com/opencollab/arpack-ng/blob/3dd2d5fe97ed40759b3b175d4a1c27ebcd9b0921/.travis.yml#L258 but not elsewhere. There are all different in many many places (apt-get install ` cmake/automake` , ` openmpi/mpich` , `  gcc\clang` if ` mpich/clang`  could be tested later, ...) : nothing to factorize here, only headache-time-wasting !?

File: .travis.yml [hunk 3]
```diff
@@ -265,41 +265,15 @@ jobs:
...
-      &&                                                                                    \
-      sudo docker cp -a ${TRAVIS_BUILD_DIR} mobydick:/tmp                                   \
-      &&                                                                                    \
-      sudo docker start -a mobydick
+    script: scripts/travis_fedora.sh setup openmpi latest
+
+  # fedora (with gcc 10 and mpich)
+  - stage: fedora
+    dist: xenial
+    script: scripts/travis_fedora.sh setup mpich rawhide
```
Comment (sylvestre) [inline .travis.yml]:
```suggestion
    script: bash -v scripts/travis_fedora.sh setup mpich rawhide
```

File: .travis.yml [hunk 4]
```diff
@@ -265,41 +265,15 @@ jobs:
...
-       cmake -DEXAMPLES=ON -DMPI=ON -DICB=ON ..                                          && \
-       export VERBOSE=1                                                                  && \
-       make all                                                                          && \
-       make test                                                                         && \
-       tail -n 300 ./Testing/Temporary/LastTest.log"                                        \
-      &&                                                                                    \
-      sudo docker cp -a ${TRAVIS_BUILD_DIR} mobydick:/tmp                                   \
-      &&                                                                                    \
-      sudo docker start -a mobydick
+    script: scripts/travis_fedora.sh setup openmpi latest
```
Comment (sylvestre) [inline .travis.yml]:
```suggestion
    script: bash -v scripts/travis_fedora.sh setup openmpi latest
```

**Comment by sylvestre** [comment]
The script is failing:
https://travis-ci.org/opencollab/arpack-ng/jobs/654128290

File: .travis.yml (hunk 2 above)
Comment (dschwoerer) [reply .travis.yml]: Well, there are ways to improve that.
The `--allow-run-as-root --oversubscribe` flags aren't needed. There are environment vars for that. That not only allows not having to patch the files, but also not having to differentiate between mpich and openmpi, as unknown env vars are ignored.
The different flags for cmake or configure can be passed via environment vars, then the overview in travis is also more useful.
It is not 100% reusable, but there could be some things that can be improved.

The [boutproject](https://github.com/boutproject/BOUT-dev/blob/master/.travis.yml) has imho a nicer interface, although they don't use docker, so things aren't that complicated. They have a [single script](https://github.com/boutproject/BOUT-dev/blob/master/.travis_script.sh), I guess for arpack-ng probably one script for each distribution would be required ...

It is more effort in the short run, but might pay off as it is easier to maintain one script rather then many copy-pasted versions :-)

**Comment by sylvestre** [comment]
This is great, many thanks!

**Comment by kloczek** [comment]
I found that even with that patch arpack still does not compiles with gcc 10.
```
make[1]: Entering directory '/home/tkloczko/rpmbuild/BUILD/arpack-ng-3.7.0/SRC'
/bin/sh ../libtool  --tag=F77   --mode=compile gfortran  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto=auto -flto-partition=none -c -o snaitr.lo snaitr.f
libtool: compile:  gfortran -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -I/usr/lib64/gfortran/modules -flto=auto -flto-partition=none -c snaitr.f  -fPIC -o .libs/snaitr.o
annobin: snaitr.f: Warning: -D_FORTIFY_SOURCE not defined
annobin: snaitr.f: Warning: -D_GLIBCXX_ASSERTIONS not defined
snaitr.f:658:35:

  376 |             call svout (logfil, 1, rnorm, ndigit,
      |                                   2
......
  658 |             call svout (logfil, 2, xtemp, ndigit,
      |                                   1
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (scalar and rank-1)
snaitr.f:729:39:

  376 |             call svout (logfil, 1, rnorm, ndigit,
      |                                   2
......
  729 |                 call svout (logfil, 2, xtemp, ndigit,
      |                                       1
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (scalar and rank-1)
make[1]: *** [Makefile:563: snaitr.lo] Error 1
make[1]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/arpack-ng-3.7.0/SRC'
make: *** [Makefile:508: all-recursive] Error 1
```
If that issue is already sorted out is it possible to make new release?

**Comment by fghoussen** [comment]
would be good to add "gcc-10 + mpich" job in CI

**Comment by dschwoerer** [comment]
There is a CI job with exactly gcc-10 and mpich:
https://github.com/opencollab/arpack-ng/blob/master/.travis.yml#L283

@kloczek This is only a fix for parpack. Arpack itself (serial version) has been fixed in #239 
This has also been backported to fedora in https://src.fedoraproject.org/rpms/arpack/pull-request/2

If the issue still persists, could you mention what version + what patches you are using?

**Comment by fghoussen** [comment]
@dschwoerer : cool, I missed you achieved that! Thanks!

**Pull request merged**